opencga-analysis/src/main/java/org/opencb/opencga/analysis/rga/RgaManager.java

```diff
@@ -1952,4 +1991,57 @@ public Preprocess setEvent(Event event) {
             return this;
         }
     }
+
+    /*
+    CACHE METHODS
+     */
+    private String generateCacheKey(String method, String studyStr, Query query, QueryOptions options) {
+        ObjectMap map = new ObjectMap()
+                .append("method", method)
+                .append("study", studyStr);
+        if (query != null) {
+            map.putAll(query);
+        }
+        if (options != null) {
+            map.putAll(options);
+        }
+        // Sort the keys
+        List<String> sortedKeys = map.keySet().stream().sorted().collect(Collectors.toList());
+        List<String> queryList = new ArrayList<>(map.size());
+        for (String key : sortedKeys) {
+            queryList.add(key + "=" + map.get(key));
+        }
+        return DigestUtils.sha256Hex(StringUtils.join(queryList, ";"));
+    }
+
+    private void cacheResults(String method, String studyStr, Query query, QueryOptions options, OpenCGAResult<?> result) {
+        if (!storageConfiguration.getRga().isCache()) {
+            // Cache is disabled
+            return;
+        }
+
+        if (cacheMap.size() > CACHE_SIZE) {
+            // Cache is already full
+            logger.warn("Query not cached. Cache is already full (size: {}).", CACHE_SIZE);
+            return;
+        }
+
+        String cacheKey = generateCacheKey(method, studyStr, query, options);
+        cacheMap.put(cacheKey, result);
```

Comment:
This method will be called from multiple threads at the same time, but the HashMap is not thread-safe.

Reply:
True, but in any case RGA is a read-only database, the worst possible scenario is one query being executed twice I think, which is the current behavior

Reply:
Don't see how the ReadOnly database applies here.

Simple scenario: Two different queries being inserted at the same time from different calls (i.e. different threads) may throw a ConcurrentModificationException, and the map itself might get corrupted.

Of course, as with most of the "race-condition" problems, it's unlikely to happen... but...

Reply:
But let's make it synced, that's easy and safer